Guide to including & accessing package data (#761)
pypa/packaging.python.org: source/guides/using-package-data.rst

There are two versions of ``importlib-resources`` available:

- `The one on PyPI`__ that is installed with ``pip install
  importlib-resources`` and imported with ``import importlib_resources`` (note
  underscore)

  .. __: https://pypi.org/project/importlib-resources/

- `The one in the Python standard library`__ starting with Python 3.7 that is
  imported with ``import importlib.resources`` (note period)

  .. __: https://docs.python.org/3/library/importlib.html
         #module-importlib.resources

Development of the PyPI version tends to be ahead of whatever's in the latest
Python version.  In particular, the new ``files()``-based API described here
was only introduced in version 1.1.0 of the PyPI project and was only added to
the Python standard library in Python 3.9.  In order to be guaranteed a version
of ``importlib-resources`` that supports this API, you should add the following
to your project's ``install_requires``::

    "importlib-resources>=1.1.0; python_version < '3.9'"

and import ``importlib-resources`` in your code as follows:

.. code-block:: python

    import sys

    if sys.version_info < (3, 9):
        import importlib_resources
    else:

>         import importlib.resources as importlib_resources

Reviewer: Why not use the typical try-except idiom?

Author: Because `try: import importlib.resources` would succeed on Python 3.7 and 3.8, but we don't want to use `importlib.resources` there, as it doesn't have the `files()` function.  If you're suggesting `try: import importlib_resources` instead, that's really a stylistic decision.

Reviewer: Good point. This should be noted explicitly...

Author: Updated.

The ``importlib-resources`` API
-------------------------------

To access a package data file in your project, start by calling
``importlib_resources.files()`` on the name of your package:

.. code-block:: python

    pkg = importlib_resources.files("packagename")
    # The argument can optionally refer to a subpackage in the form
    # "packagename.subpackage".

> This gives you an object with a subset of `pathlib.Path`'s methods for

Reviewer: I believe that no explicit role for single backticks is set and so you should use it explicitly like
```suggestion
This gives you an object with a subset of :py:class:`pathlib.Path`'s methods for
```

Author: Updated.

When creating a Python project, you may want to include a number of non-Python
files in the project that the code can then access at runtime, such as
templates, images, and data.  These files are called *package data*, and they
are included in & accessed from your project as follows.

Including package data
======================

The first requirement for being able to include package data in your package is
for the files to be located somewhere inside the directory that is your
project's :term:`import package <Import Package>`.  If your project is a
single-file ``.py`` module, you must convert it to a directory/package layout
before you can store package data in it.

Once your package data is inside the import package, you must tell setuptools
to include the files in both the :term:`wheel <Wheel>` (so that the files will
be installed when your package is installed) and :term:`sdist <Source
Distribution (or "sdist")>` (so that the files can be included in the wheel
when building from an sdist).  There are two ways to do this: a blanket
"include all" method and a fine-grained method.

Blanket Method
--------------

There are two steps to this method:

> 1. In your :file:`setup.py`, pass ``include_package_data=True`` to the

Reviewer: Hm... Should this mention that this file may also be `setup.cfg`?

Author: Seeing as nothing else in packaging.python.org references putting configuration in `setup.cfg` (I think), this would probably be more confusing than helpful.  People who already know configuration can be placed in `setup.cfg` should be able to figure out that `include_package_data` can be placed there as well.

   ``setup()`` function.  This tells setuptools that any non-Python files found
   inside your import package that are included in the sdist should also be
   included in the wheel.

2. Create or edit your project's :file:`MANIFEST.in` file with the proper
   commands so that all of your package data files are included in the
   project's sdists.  See :ref:`Using MANIFEST.in` for information on how to do
   this.

   In the simplest case — in which you want all non-Python files in your import
   package to be treated as package data — you can simply include your entire
   import package in your sdist with the :file:`MANIFEST.in` command
   ":samp:`graft {packagename}`" (or ":samp:`graft src`" if you're using a

>    ``src/`` layout).  If you use this command, you should also follow it with

Reviewer: I think, it'd be great to link src layout docs here

Author: What src layout docs?

Reviewer: I was thinking about https://setuptools.readthedocs.io/en/latest/setuptools.html#configuring-setup-using-setup-cfg-files

Author: That's about `setup.cfg`, not `src/`.

Reviewer: Ah, sorry, I meant https://setuptools.readthedocs.io/en/latest/setuptools.html#using-a-src-layout

Author: That's just about how to edit `setup.cfg` for a `src/` layout.  I don't think it's worth linking here.

Reviewer: How about https://hynek.me/articles/testing-packaging/

Author: Done.

   ":samp:`global-exclude *.py[cod]`" so that compiled Python bytecode files
   are not included.

Fine-Grained Method
-------------------

The fine-grained method allows you to specify package data file patterns on a
per-package basis through the ``package_data`` argument to ``setup()``.
``package_data`` must be a :py:class:`dict` mapping import package or

> subpackage names (or the empty string, to specify all packages & subpackages)

Reviewer: This would be an asterisk for `setup.cfg` IIRC

Author: A passage on performing this configuration in `setup.cfg` has been added.

to lists of glob patterns indicating which files within those packages to
include in the sdist and wheel.

A sample ``package_data`` looks like this:

.. code-block:: python

    package_data={
        # If any package or subpackage contains *.txt or *.rst files, include
        # them:
        "": ["*.txt", "*.rst"],
        # Include any *.msg files found in the "hello" package (but not in its
        # subpackages):
        "hello": ["*.msg"],
        # Include any *.csv files found in the "hello.utils" package:
        "hello.utils": ["*.csv"],
        # Include any *.dat files found in the "data" subdirectory of the
        # "mypkg" package:
        "mypkg": ["data/*.dat"],
    }

> Note that glob patterns only select files located directly within the given

Reviewer: Maybe use something like this? 
```suggestion
.. tip::

   Note that glob patterns only select files located directly within the given
```

package (or in the given subdirectory of the package, if the pattern includes a
directory path); e.g., ``"hello": ["*.msg"]`` selects ``*.msg`` files in the
``hello`` package but not in any of its subpackages.  To select files in
subpackages, you must either include an entry for each subpackage or else use
the empty string key to specify a pattern for all packages & subpackages.

If a pattern contains any directory components, the forward slash (``/``) must
be used as the directory separator, even on Windows.

If a package data file is located in a directory that does not have an

> :file:`__init__.py` file (say, a ``data/`` directory inside

Reviewer: Is this also true when `find_namespace_packages()` is used?

Author: I believe so.  I'm pretty sure `find_namespace_packages()` has no interaction with `package_data`.

.. warning::

    If you use both ``include_package_data`` and ``package_data``, files
    specified with ``package_data`` will not be automatically included in
    sdists; you must instead list them in your :file:`MANIFEST.in`.

Excluding Files
---------------

The ``exclude_package_data`` argument to ``setup()`` can be used in conjunction
with either of the above methods to prevent one or more files from being
treated as package data.  ``exclude_package_data`` takes a :py:class:`dict`
with the same structure as ``package_data``, and any matched files are excluded
from wheels.  Matched files are also excluded from sdists if they are not
already matched by the project's :file:`MANIFEST.in`.

Accessing package data
======================

There have been multiple ways to access package data over the years, from
|pkg_resources' ResourceManager API|__ to |pkgutil.get_data()|__, but the most
recent and currently-recommended way is with the `importlib-resources`__
package.

.. |pkg_resources' ResourceManager API| replace:: ``pkg_resources``' ``ResourceManager`` API
.. __: https://setuptools.readthedocs.io/en/latest/pkg_resources.html
       #resourcemanager-api

> .. |pkgutil.get_data()| replace:: ``pkgutil.get_data()``

Reviewer: Isn't this linkable via intersphinx?

Author: Linked (though it goes to the Python 2 docs for some reason).

Reviewer: You may explicitly prefix the ref with the intersphinx identifier to force linking a certain docs site

Author: Done.

.. __: https://docs.python.org/3/library/pkgutil.html#pkgutil.get_data

.. __: http://importlib-resources.readthedocs.io

Installing & importing ``importlib-resources``
----------------------------------------------

There are two versions of ``importlib-resources`` available:

- `The one on PyPI`__ that is installed with ``pip install
  importlib-resources`` and imported with ``import importlib_resources`` (note
  underscore)

  .. __: https://pypi.org/project/importlib-resources/

> - `The one in the Python standard library`__ starting with Python 3.7 that is

Reviewer: Since I've upgraded the intersphinx links, this and some other entries should be linkable through normal roles/refs

Author: I tried rewriting this link in the following ways:

```
:py:mod:`The one in the Python standard library <importlib.resources>`
:ref:`The one in the Python standard library <python:importlib.resources>`
:ref:`The one in the Python standard library <python:module:importlib.resources>`
```

The first link was styled wrong.  The other two produced "undefined label" warnings, treated as errors.

Reviewer 2: I [suggested a change](https://github.com/pypa/packaging.python.org/pull/761/files#r477557831) to handle the linking problem you're having.

version control that you want to exclude from sdists or wheels), though you
still need to set ``include_package_data`` to ``True`` for files in your import
package directory to be included in wheels.

.. _setuptools_scm: https://github.com/pypa/setuptools_scm

Accessing package data
======================

There have been multiple ways to access package data over the years, from
|pkg_resources' ResourceManager API|__ to :py:func:`pkgutil.get_data()`, but
the most recent and currently-recommended way is with the
`importlib-resources`__ package.

.. |pkg_resources' ResourceManager API| replace:: ``pkg_resources``' ``ResourceManager`` API
.. __: https://setuptools.readthedocs.io/en/latest/pkg_resources.html
       #resourcemanager-api

.. __: http://importlib-resources.readthedocs.io

Installing & importing ``importlib-resources``
----------------------------------------------

There are two versions of ``importlib-resources`` available:

- `The one on PyPI`__ that is installed with ``pip install
  importlib-resources`` and imported with ``import importlib_resources`` (note
  underscore)

  .. __: https://pypi.org/project/importlib-resources/

> - `The one in the Python standard library`__ starting with Python 3.7 that is
>   imported with ``import importlib.resources`` (note period)

Reviewer 2: ```suggestion
- The Python standard library's starting with Python 3.7 that is
  imported with ``import`` :py:mod:`python:importlib.resrouces` (note period)
```

        import importlib_resources
    else:
        # importlib.resources has files(), so use that:
        import importlib.resources as importlib_resources

The ``importlib-resources`` API
-------------------------------

To access a package data file in your project, start by calling
``importlib_resources.files()`` on the name of your package:

.. code-block:: python

    pkg = importlib_resources.files("packagename")
    # The argument can optionally refer to a subpackage in the form
    # "packagename.subpackage".

This gives you an object with a subset of :py:class:`pathlib.Path`'s methods
for traversing package data files.  To refer to a :file:`data.csv` file in a
``data/`` directory in your package, write:

.. code-block:: python

    pkg_data_file = pkg / "data" / "data.csv"

So now that we've got a reference to the package data file, how do we get
anything out of it?

- To open the file for reading, call the ``open()`` method:

  .. code-block:: python

    with pkg_data_file.open() as fp:
        # Do things with fp

- To get the file's contents as :py:class:`bytes`, call the ``read_bytes()``
  method:

  .. code-block:: python

>     b = pkg_data_file.read_bytes()

Reviewer 2: This can also be done without the `files` API using [`importlib.resources.read_binary`](https://docs.python.org/3/library/importlib.html#importlib.resources.read_binary)

Reviewer 3: This API, while still supported, is legacy and discouraged.

- To get the file's contents as a :py:class:`str`, call the ``read_text()``
  method, optionally with an ``encoding`` argument:

  .. code-block:: python

>     s = pkg_data_file.read_text(encoding="utf-8")

Reviewer 2: This can also be done without the `files` API using [`importlib.resources.read_text`](https://docs.python.org/3/library/importlib.html#importlib.resources.read_text)

Reviewer 3: discouraged

> - To get the path to the file, call ``importlib_resources.as_file()`` on it and

Reviewer 2: This can also be done without the `files` API using [`importlib.resources.path`](https://docs.python.org/3/library/importlib.html#importlib.resources.path)

Reviewer 3: discouraged

  use the return value as a context manager:

  .. code-block:: python

    with importlib_resources.as_file(pkg_data_file) as path:
        # Do things with the pathlib.Path object that is `path`

  The use of context managers allows ``importlib-resources`` to support
  packages stored in zipfiles; when a path is requested for a package data file
  in a zipfile, the library can extract the file to a temporary location at the
  start of the ``with`` block and remove it at the end of the block.

- To iterate through a directory (either a package or a non-package directory),
  use the ``iterdir()`` method.  You can test whether a resource is a directory
  or a file with the ``is_dir()`` and ``is_file()`` methods, and you can get a
  resource's basename via the ``name`` property:

  .. code-block:: python

>     for entry in (pkg / "data").iterdir():

Reviewer 2: This can be done without the `files` API using [`importlib.resources.contents`](https://docs.python.org/3/library/importlib.html#importlib.resources.contents)

Reviewer 3: discouraged

        if entry.is_dir():
            print(entry.name, "DIR")

>         else:

Reviewer 2: This can be done without the `files` API using [`importlib.resources.is_resource`](https://docs.python.org/3/library/importlib.html#importlib.resources.is_resource)

Reviewer 3: discouraged

.. _`Using package data`:

==================================
Including & accessing package data
==================================

When creating a Python project, you may want to include a number of non-Python
files in the project that the code can then access at runtime, such as
templates, images, and data.  These files are called *package data*, and they
are included in & accessed from your project as follows.

Including package data
======================

.. note::

    The following instructions only pertain to projects built with setuptools.
    Users of flit, Poetry, and other tools must consult the respective

>     backend's documentation to see how to include package data.

Reviewer 3: I would not recommend documenting setuptools-specific behaviors in the packaging docs. These docs should present general packaging concerns and then defer to the setuptools docs for setuptools-specific implementation details (and other tools for their related details). By attempting to document here, you create a duplication of effort with what the setuptools docs attempt to provide. Since we're already short-staffed on maintenance support, I'm reluctant to have more documentation that might become out of sync with the implementation.

The first requirement for being able to include package data in your package is
for the files to be located somewhere inside the directory that is your

> project's :term:`import package <Import Package>`.  If your project is a

Reviewer 3: Here or above, I think it would be useful to link to [Python's documentation for packages](https://docs.python.org/3/tutorial/modules.html#packages) and then describe the relationship between a Python Package (I wouldn't call it an Import Package) and a Distribution Package and indicate which form of 'package' is used in the documentation.

single-file ``.py`` module, you must convert it to a directory/package layout
before you can store package data in it.

Once your package data is inside the import package, you must tell setuptools
to include the files in both the :term:`wheel <Wheel>` (so that the files will
be installed when your package is installed) and :term:`sdist <Source

> Distribution (or "sdist")>` (so that the files can be included in the wheel

Reviewer 3: Here, I would focus on the different between a source distribution (aka sdist, i.e. a copy of the source suitable for another person to build, so includes the packaging scaffolding and testing and other ancillary features) and a built distribution (aka bdist or wheel, which is a distribution that's built and ready for installation). The concepts for a built distribution also apply to a package that's built and directly installed (from source with no intermediary built distribution, i.e. `setup.py install`), although this workflow is largely discouraged.

when building from an sdist).  There are two main ways to do this: a blanket
"include all" method (in which package data is configured through the
:file:`MANIFEST.in` file) and a fine-grained method (in which package data is
configured in :file:`setup.py` or :file:`setup.cfg` one package & subpackage at
a time).

Blanket method
--------------

There are two steps to this method:

1. In your :file:`setup.py`, pass ``include_package_data=True`` to the
   ``setup()`` function.   If you are placing your project configuration in
   :file:`setup.cfg` instead, set ``include_package_data = True`` in the
   ``[options]`` section of :file:`setup.cfg`.

   Doing this tells setuptools that any non-Python files found inside your
   import package that are included in the sdist should also be included in the
   wheel.

> 2. Create or edit your project's :file:`MANIFEST.in` file with the proper

Reviewer 3: There are other ways to ensure the files are included in the sdist. My preferred way and the way I recommend is to use the [setuptools_scm project](https://pypi.org/project/setuptools-scm) to discover all sources from SCM metadata. This approach is simple and maintenance free.

:file:`__init__.py` file (say, a ``data/`` directory inside
``package.subpackage``), that directory does not count as a package, and the
file must be listed in ``package_data`` in the form
:samp:`"package.subpackage": ["data/{pattern}"]`.

.. warning::

    If you use both ``include_package_data`` and ``package_data``, files
    specified with ``package_data`` will not be automatically included in
    sdists; you must instead list them in your :file:`MANIFEST.in`.

Excluding files
---------------

The ``exclude_package_data`` argument to ``setup()`` can be used in conjunction
with either of the above methods to prevent one or more files from being
treated as package data.  ``exclude_package_data`` takes a :py:class:`dict`
with the same structure as ``package_data``, and any matched files are excluded
from wheels.  Matched files are also excluded from sdists if they are not
already matched by the project's :file:`MANIFEST.in`.

In a :file:`setup.cfg`, ``exclude_package_data`` becomes an
``[options.exclude_package_data]`` section whose contents have the same
structure as ``[options.package_data]``.

Including files via setuptools plugins
--------------------------------------

> As an alternative to the above methods, you can use a plugin for setuptools

Reviewer 3: The use of plugins is only an alternative to the manifest template (MANIFEST.in). I instead would suggest breaking down the problem into two phases: file discovery for sources (handled by plugin or manifest template) and file inclusion for installation (and wheels), handled by 'package_data' or 'include_package_data' directive.

that automatically recognizes & includes package data in sdists & wheels,
usually based on what files in the project directory are under verson control.
One example of such a plugin is setuptools_scm_, which automatically finds all
files under version control in a Git or Mercurial repository and augments the

> project's :file:`MANIFEST.in` (if any) with the found files.  This eliminates

Reviewer 3: This phrasing may give the reader the impression that the file is somehow altered. I'd instead say that it affects the setuptools source file discovery mechanism, augmenting the list of files discovered (in addition to those found through the manifest template).

the need to write a :file:`MANIFEST.in` manually (unless there are files under
version control that you want to exclude from sdists or wheels), though you
still need to set ``include_package_data`` to ``True`` for files in your import
package directory to be included in wheels.

.. _setuptools_scm: https://github.com/pypa/setuptools_scm

Accessing package data
======================

There have been multiple ways to access package data over the years, from
|pkg_resources' ResourceManager API|__ to :py:func:`python:pkgutil.get_data()`,
but the most recent and currently-recommended way is with the
`importlib-resources`__ package.

.. |pkg_resources' ResourceManager API| replace:: ``pkg_resources``' ``ResourceManager`` API
.. __: https://setuptools.readthedocs.io/en/latest/pkg_resources.html
       #resourcemanager-api

.. __: http://importlib-resources.readthedocs.io

Installing & importing ``importlib-resources``
----------------------------------------------

> There are two versions of ``importlib-resources`` available:

Reviewer 3: Again, here the documentation is repeating a lot of information that's authoritative elsewhere. It may be more sustainable to give a high-level introduction and leave the details to the projects to document (and maybe contribute there).

- `The one on PyPI`__ that is installed with ``pip install
  importlib-resources`` and imported with ``import importlib_resources`` (note
  underscore)

  .. __: https://pypi.org/project/importlib-resources/

- `The one in the Python standard library`__ starting with Python 3.7 that is
  imported with ``import importlib.resources`` (note period)

  .. __: https://docs.python.org/3/library/importlib.html
         #module-importlib.resources

Development of the PyPI version tends to be ahead of whatever's in the latest

> Python version.  In particular, the new ``files()``-based API described here

Reviewer 3: This API is recommended and intended to supersede the other API functions.

was only introduced in version 1.1.0 of the PyPI project and was only added to
the Python standard library in Python 3.9.  In order to be guaranteed a version
of ``importlib-resources`` that supports this API, you should add the following
to your project's ``install_requires``::

    "importlib-resources>=1.1.0; python_version < '3.9'"

and import ``importlib-resources`` in your code as follows:

.. code-block:: python

    import sys

    if sys.version_info < (3, 9):
        # importlib.resources either doesn't exist or lacks the files()
        # function, so use the PyPI version:

>         import importlib_resources
>     else:
>         # importlib.resources has files(), so use that:
>         import importlib.resources as importlib_resources

Reviewer 3: I'd recommend use a `from importlib` import for brevity.

```suggestion
        import importlib_resources as resources
    else:
        # importlib.resources has files(), so use that:
        from importlib import resources
```

The ``importlib-resources`` API
-------------------------------

To access a package data file in your project, start by calling
``importlib_resources.files()`` on the name of your package:

.. code-block:: python

    pkg = importlib_resources.files("packagename")
    # The argument can optionally refer to a subpackage in the form
    # "packagename.subpackage".

> This gives you an object with a subset of :py:class:`pathlib.Path`'s methods

Reviewer 3: This object now has a formal Protocol, [Traversable](https://github.com/python/importlib_resources/blob/eb0554035b499e4a5349dcc56a2bda51835f4fde/importlib_resources/abc.py#L61).
